blockntfns/manager.go

```diff
@@ -0,0 +1,348 @@
+				case <-sub.quit:
+					return
+				case <-m.quit:
+					return
+				}
+			case <-sub.quit:
+				return
+			case <-m.quit:
+				return
+			}
+		}
+	}()
+
+	// Now, we can deliver the notification to the subscription handler.
+	select {
+	case m.newSubscriptions <- sub:
+	case <-m.quit:
+		sub.ntfnQueue.Stop()
+		return nil, ErrSubscriptionManagerStopped
+	}
+
+	// It's possible that the registration failed if we were unable to
+	// deliver the backlog of notifications, so we'll make sure to handle
+	// the error.
+	select {
+	case err := <-sub.errChan:
+		if err != nil {
+			sub.ntfnQueue.Stop()
+			return nil, err
+		}
+	case <-m.quit:
+		sub.ntfnQueue.Stop()
+		return nil, ErrSubscriptionManagerStopped
+	}
+
+	// Finally, we can return to the client with its new subscription
+	// successfully registered.
+	return &Subscription{
+		Notifications: sub.ntfnChan,
+		Cancel: func() {
+			m.cancelSubscription(sub)
+		},
+	}, nil
+}
+
+// handleNewSubscription handles a request to create a new block subscription.
+func (m *SubscriptionManager) handleNewSubscription(sub *newSubscription) error {
+	log.Infof("Registering block subscription: id=%d", sub.id)
+
+	// We'll start by retrieving a backlog of notifications from the
+	// client's best height.
+	blocks, currentHeight, err := m.ntfnSource.NotificationsSinceHeight(
+		sub.bestHeight,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve blocks since height=%d: "+
+			"%v", sub.bestHeight, err)
+	}
+
+	// We'll then attempt to deliver these notifications.
+	log.Debugf("Delivering backlog of block notifications: id=%d, "+
+		"start_height=%d, end_height=%d", sub.id, sub.bestHeight,
+		currentHeight)
+
+	for _, block := range blocks {
+		m.notifySubscriber(sub, block)
+	}
+
+	// With the notifications delivered, we can keep track of the new client
+	// internally in order to deliver new block notifications about the
+	// chain.
+	m.subscribers[sub.id] = sub
+
+	return nil
+}
+
+// cancelSubscription sends a request to the subscription handler to cancel an
+// existing subscription.
+func (m *SubscriptionManager) cancelSubscription(sub *newSubscription) {
+	select {
+	case m.cancelSubscriptions <- &cancelSubscription{sub.id}:
+	case <-m.quit:
+	}
+}
+
+// handleCancelSubscription handles a request to cancel an existing
+// subscription.
+func (m *SubscriptionManager) handleCancelSubscription(msg *cancelSubscription) {
+	// First, we'll attempt to look up an existing susbcriber with the given
+	// ID.
+	sub, ok := m.subscribers[msg.id]
+	if !ok {
+		return
+	}
+
+	log.Infof("Canceling block subscription: id=%d", msg.id)
+
+	// If there is one, we'll stop their internal queue to no longer deliver
+	// notifications to them.
+	delete(m.subscribers, msg.id)
+	sub.cancel()
+}
+
+// notifySubscribers notifies all currently active subscribers about the block.
+func (m *SubscriptionManager) notifySubscribers(ntfn BlockNtfn) {
+	log.Infof("Notifying %v", ntfn)
```

Comment:
This should be demoted to trace, otherwise IBD will be _very_ spammy (it is rn with this).

Reply:
Fixed.